spinalcordtoolbox/utils/csi_filter.py

```diff
@@ -0,0 +1,100 @@
+"""
+Color code removal filter, implemented as a codec
+
+Should be usable for writing to a file, for example:
+
+>>> from spinalcordtoolbox.utils import csi_filter
+>>> csi_filter.register_codec()
+>>> with open('some_file', 'w', encoding='csi-filter') as some_file:
+...     some_file.write('Some \x1b[1mBRIGHT\x1b[m text\n')
+24
+>>> open('some_file', 'rb').read()
+'Some BRIGHT text\n'
+
+Copyright (c) 2024 Polytechnique Montreal <www.neuro.polymtl.ca>
+License: see the file LICENSE
+"""
+
+import codecs
+import re
+
+# a complete CSI escape sequence (cursor movements, color codes, etc.)
+csi_escape = re.compile(
+    r"""
+    \x1b \[  # the control sequence introducer, ESC [
+    [0-?]*   # optional parameter bytes, between 0x30 and 0x3f
+    [ -/]*   # optional intermediate bytes, between 0x20 and 0x2f
+    [@-~]    # required final byte, between 0x40 and 0x7e
+    """,
+    re.VERBOSE,
+)
+
+# The last part of the input which *might* start a CSI escape sequence.
+# This matches every possible *incomplete* CSI escape sequence, including
+# the empty string for convenience (so that the regex always matches).
+csi_suffix = re.compile(
+    r"""
+    \Z
+    | \x1b \Z
+    | \x1b \[ [0-?]* [ -/]* \Z
```

Comment:
TIL (possibly not for the first time) that in Python regex, `$` can match at places other than the end of the string, even without `re.MULTILINE`! Specifically, it can match _one character before the end of string_ as well, if the last character is a newline. But in this case, I really wanted the end of string.

I noticed this because the little example I wrote at the top of the file didn't work at first: the newline character at the end got swallowed by the encoder.